File: src/bin/particle.rs
use poincare_maps::*;
use std::path::PathBuf;

fn main() {
    let path = PathBuf::from("./data.nc");
    let qfactor = Qfactor::from_dataset(&path, "akima").unwrap();
    let bfield = Bfield::from_dataset(&path, "bicubic").unwrap();
    let current = Current::from_dataset(&path, "akima").unwrap();

    // Normal passing particle
    let initial = InitialConditions {
        t0: 0.0,
        theta0: 3.14,
        psip0: 0.05,
        rho0: 0.05,
        zeta0: 0.1,
        mu: 1e-4,
    };
    let mut particle = Particle::new(initial);
    particle
        .run_ode(&qfactor, &bfield, &current, (0.0, 200.0), 5000)
        .unwrap();
    dbg!(&particle);
}

File: scripts/particle.py
#!/usr/bin/env -S uv run --script --active
# /// script
# requires-python = ">=3.14"
# dependencies = [
#   "pyncare"
#   "matplotlib"
# ]
# ///
import matplotlib
import pyncare as pc

matplotlib.use("gtk3agg")

qfactor = pc.Qfactor("./data.nc", "akima")
current = pc.Current("./data.nc", "akima")
bfield = pc.Bfield("./data.nc", "bicubic")
per = pc.Perturbation(
    [
        pc.Harmonic("./data.nc", "akima", m=1, n=2, phase=0),
        pc.Harmonic("./data.nc", "akima", m=3, n=2, phase=0),
    ]
)

initial = pc.InitialConditions(
    t0=0,
    theta0=3.14,
    psip0=0.5 * qfactor.psip_wall,
    rho0=0.001,
    zeta0=0.0,
    mu=0,
)

particle = pc.Particle(initial)

particle.integrate(
    qfactor=qfactor,
    current=current,
    bfield=bfield,
    per=per,
    t_eval=[0, 20000],
)
print(particle)

pc.orbit_plot(particle)
